LAN78xx VLAN checksum fix

6by9 opened the pull request:
Implementing VLAN header stripping in the LAN78xx appears to make the checksum issues go away in my test system.

This patch also exposes VLAN filtering - I'm not sure that really adds that much on a sensibly configured switch, but the functionality was 90% there anyway so it seemed worth completing. (That could be split out into a second patch if you want it that way).

With these patches it is still possible to execute ```ethtool -K eth0 rx-vlan-hw-parse off``` and switch into the old bad state of incorrect checksums. Do we want to try and cover that one off, or leave it that the defaults work? It's a couple of extra conditionals in lan78xx_rx_csum_offload to keep part of the workaround in place if VLAN stripping isn't enabled, but that code is going to be executed for every received frame therefore it wants to be as efficient as possible.

6by9 committed:
net: lan78xx: Add support for VLAN filtering.

HW_VLAN_CTAG_FILTER was partially implemented, but not fully to Linux.
Complete the implementation of this.

See #2458.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
net: lan78xx: Reduce s/w csum check on VLANs

With HW_VLAN_CTAG_RX enabled we don't observe the checksum
issue, so amend the workaround to only drop back to s/w
checksums if VLAN offload is disabled.

See #2458.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
Tag to #2458

popcornmix commented:
Looks okay to me.
I'm not too concerned about being able to enable the buggy behaviour with ethtool - perhaps upstream will have a preference?

pelwell commented:
I suggested falling back to SW checksum in the case that the HW parsing is disabled.

6by9 committed:
net: lan78xx: Add support for VLAN tag stripping.

The chip supports stripping the VLAN tag and reporting it
in metadata. Implement this as it also appears to solve the
issues observed in checksum computation.

See #2458.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
Updated, but it's a cherry-pick from another branch so I'm currently building it to ensure it is good.

6by9 commented:
OK, looking fine to me.
Sorry, forgot to tag the original issue in the commit text when I rewrote it :-(

pelwell commented:
I can't rewrite the commit message without squashing, so you'll have to fix it on your end.

6by9 commented:
OK, patches reworded and pushed again. No changes to the code on the latest push.
